File: apps/web/server/middleware/security.ts
import { appendResponseHeader, eventHandler } from 'h3';
import { env } from '../../app/lib/config/env';

/**
 * Security headers middleware for TanStack Start.
 * Ported from Next.js proxy.ts.
 */
export default eventHandler((event) => {
  const isProduction = env.NODE_ENV === 'production';
  const scriptSrc = isProduction
    ? [
        "script-src 'self'",
        'https://js.stripe.com',
        'https://va.vercel-scripts.com',
        'https://vercel.live',
        'blob:',
      ].join(' ')
    : "script-src 'self' 'unsafe-eval' 'unsafe-inline' https: blob:";
  const styleSrc = isProduction ? "style-src 'self'" : "style-src 'self' 'unsafe-inline'";
  const connectSrc = isProduction
    ? [
        "connect-src 'self'",
        'https://taskforceai.chat',
        'https://*.taskforceai.chat',
        'https://api.vercel.ai',
        'https://vitals.vercel-insights.com',
        'https://*.sentry.io',
        'https://*.ingest.sentry.io',
        'https://*.public.blob.vercel-storage.com',
        'wss://*.taskforceai.chat',
        'ipc://localhost',
      ].join(' ')
    : "connect-src 'self' https: wss: ipc://localhost";
  const frameSrc = isProduction
    ? "frame-src 'self' https://js.stripe.com https://*.stripe.com"
    : "frame-src 'self' https: data:";

  // Content Security Policy
  const cspDirectives = [
    "default-src 'self'",
    scriptSrc,
    styleSrc,
    "img-src 'self' data: blob: https:",
    "font-src 'self' data:",
    connectSrc,
    frameSrc,
    "worker-src 'self' blob:",
    "frame-ancestors 'none'",
    "base-uri 'self'",
    "form-action 'self'",
  ];

  appendResponseHeader(event, 'Content-Security-Policy', cspDirectives.join('; '));

  // Strict Transport Security (HSTS) - production only
  if (isProduction) {
    appendResponseHeader(
      event,
      'Strict-Transport-Security',
      'max-age=63072000; includeSubDomains; preload'
    );
  }

  // Prevent clickjacking attacks
  appendResponseHeader(event, 'X-Frame-Options', 'DENY');

  // Prevent MIME type sniffing
  appendResponseHeader(event, 'X-Content-Type-Options', 'nosniff');

  // XSS Protection (legacy but still useful for older browsers)
  appendResponseHeader(event, 'X-XSS-Protection', '0');

  // Permissions Policy - control which browser features can be used
  appendResponseHeader(event, 'Permissions-Policy', 'camera=(), microphone=(self), geolocation=()');
});

File: packages/adapters/pkg/handler/security.go
package handler

import (
	"net/http"
	"os"
	"strings"
)

// WithRecovery is middleware that recovers from panics and logs the error.
func WithRecovery(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if err := recover(); err != nil {
				getPanicReporter().ReportRequestPanic(r, err)

				GetLogger().Error("Panic recovered", map[string]any{
					"error": err,
					"path":  r.URL.Path,
				})
				JSONError(w, http.StatusInternalServerError, "Internal server error")
			}
		}()
		next(w, r)
	}
}

// WithCSRF is middleware that validates a CSRF token for state-changing requests.
// Implements the Double Submit Cookie pattern for browser-based clients.
func WithCSRF(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		// 1. Exempt safe methods
		if r.Method == http.MethodGet || r.Method == http.MethodHead || r.Method == http.MethodOptions || r.Method == http.MethodTrace {
			next(w, r)
			return
		}

		// 2. Known webhook endpoints are exempt
		path := r.URL.Path
		if isWebhookCSRFExemptPath(path) {
			next(w, r)
			return
		}
		if isLocalTestLoginCSRFExemptPath(path) {
			next(w, r)
			return
		}
		if isNativeOAuthTokenExchangePath(path) {
			next(w, r)
			return
		}

		// 3. Client Check: Non-browser clients (TUI, Mobile, Desktop) are exempt from CSRF
		// because they don't automatically send session cookies and are not vulnerable to CSRF attacks.
		// Header-only auth is exempt only when no session auth cookie is present.
		ua := r.Header.Get("User-Agent")
		hasSessionCookie := hasSessionAuthCookie(r)
		authHeader := strings.TrimSpace(r.Header.Get("Authorization"))
		hasBearerToken := strings.HasPrefix(strings.ToLower(authHeader), "bearer ") && len(strings.TrimSpace(authHeader[len("Bearer "):])) > 0
		hasAPIKey := strings.TrimSpace(r.Header.Get("x-api-key")) != ""
		hasHeaderOnlyAuth := (hasBearerToken || hasAPIKey) && !hasSessionCookie
		isNonBrowser := strings.Contains(ua, "taskforceai-cli") ||
			strings.Contains(ua, "TaskForceAI-Desktop") ||
			strings.Contains(ua, "TaskForceAI-Mobile") ||
			hasHeaderOnlyAuth

		if isNonBrowser {
			next(w, r)
			return
		}

		// 4. Browser-based CSRF Validation (Double Submit Cookie)
		csrfToken := r.Header.Get("X-CSRF-Token")
		if csrfToken == "" {
			GetLogger().Warn("CSRF token missing", map[string]any{
				"path":              r.URL.Path,
				"method":            r.Method,
				"ua":                ua,
				"hasSessionCookie":  hasSessionCookie,
				"hasBearerToken":    hasBearerToken,
				"hasAPIKey":         hasAPIKey,
				"hasHeaderOnlyAuth": hasHeaderOnlyAuth,
			})
			JSONError(w, http.StatusForbidden, "CSRF token missing")
			return
		}

		cookie, err := r.Cookie("csrf_token")
		if err != nil {
			GetLogger().Warn("CSRF cookie missing", map[string]any{
				"path":              r.URL.Path,
				"method":            r.Method,
				"ua":                ua,
				"hasSessionCookie":  hasSessionCookie,
				"hasBearerToken":    hasBearerToken,
				"hasAPIKey":         hasAPIKey,
				"hasHeaderOnlyAuth": hasHeaderOnlyAuth,
			})
			JSONError(w, http.StatusForbidden, "CSRF cookie missing")
			return
		}

		if cookie.Value != csrfToken {
			GetLogger().Warn("CSRF token mismatch", csrfMismatchMetadata(r.URL.Path, csrfToken, cookie.Value))
			JSONError(w, http.StatusForbidden, "CSRF token mismatch")
			return
		}

		next(w, r)
	}
}

func hasSessionAuthCookie(r *http.Request) bool {
	cookieNames := []string{"__Secure-session_token", "session_token"}
	for _, name := range cookieNames {
		if cookie, err := r.Cookie(name); err == nil && strings.TrimSpace(cookie.Value) != "" {
			return true
		}
	}

	return false
}

func isWebhookCSRFExemptPath(path string) bool {
	return strings.HasPrefix(path, "/api/v1/auth/webhooks/") ||
		path == "/api/v1/auth/webhooks" ||
		path == "/api/inngest" ||
		strings.HasPrefix(path, "/api/inngest/") ||
		path == "/api/v1/payments/webhook" ||
		path == "/api/v1/payments/webhook/revenuecat"
}

func isLocalTestLoginCSRFExemptPath(path string) bool {
	if path != "/api/v1/auth/test-login" {
		return false
	}
	if IsProductionEnv() {
		return false
	}
	return isGoTestBinary() &&
		strings.EqualFold(strings.TrimSpace(os.Getenv("GO_ENV")), "test") &&
		strings.EqualFold(strings.TrimSpace(os.Getenv("ENABLE_TEST_LOGIN")), "true")
}

func isGoTestBinary() bool {
	return strings.HasSuffix(os.Args[0], ".test")
}

func isNativeOAuthTokenExchangePath(path string) bool {
	return path == "/api/v1/auth/apple" || path == "/api/v1/auth/google"
}

func csrfMismatchMetadata(path, headerToken, cookieToken string) map[string]any {
	return map[string]any{
		"path":             path,
		"headerPresent":    headerToken != "",
		"cookiePresent":    cookieToken != "",
		"headerLength":     len(headerToken),
		"cookieLength":     len(cookieToken),
		"valuesEqual":      headerToken == cookieToken,
		"mismatchDetected": true,
	}
}

// WithSecurityHeaders adds standard security headers to the response.
func WithSecurityHeaders(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-Content-Type-Options", "nosniff")
		w.Header().Set("X-Frame-Options", "DENY")
		w.Header().Set("X-XSS-Protection", "0")
		w.Header().Set("Referrer-Policy", "strict-origin-when-cross-origin")

		// Content Security Policy - Scope to HTML only
		accept := r.Header.Get("Accept")
		if strings.Contains(accept, "text/html") {
			// Relaxed policy for the web app
			w.Header().Set("Content-Security-Policy", "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://va.vercel-scripts.com; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; img-src 'self' data: blob: https:; connect-src 'self' https:; frame-ancestors 'none';")
		} else {
			// For non-HTML (API/JSON), use a highly restrictive policy if any
			w.Header().Set("Content-Security-Policy", "default-src 'none'; frame-ancestors 'none';")
		}

		// HSTS (1 year)
		if IsProductionEnv() {
			w.Header().Set("Strict-Transport-Security", "max-age=31536000; includeSubDomains; preload")
		}

		next(w, r)
	}
}
